net_udp.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package procfs
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+)
+
+type (
+	// NetUDPLine is a line parsed from /proc/net/udp
+	// For the proc file format details, see https://linux.die.net/man/5/proc
+	NetUDPLine struct {
+		TxQueue uint64
+		RxQueue uint64
+	}
+
+	NetUDP struct {
+		TxQueueLength uint64
+		RxQueueLength uint64
+		UsedSockets   uint64
+	}
+)
+
+// NetUDP returns kernel/networking statistics for udp datagrams read from /proc/net/udp.
+func (fs FS) NetUDP() (*NetUDP, error) {
+	return newNetUDP(fs.proc.Path("net/udp"))
+}
+
+// newNetUDP creates a new NetUDP from the contents of the given file.
+func newNetUDP(file string) (*NetUDP, error) {
+	f, err := os.Open(file)
```

Comment:
We're trying to make use of the procfs `utli.ReadFileNoStat()`, in order to read this data more atomically from the kernel.

But, look at this file, it seems like it could be very big in a real-world environment. Do you have any data on what size this file can grow to?

Reply:
We have a hard limit on our hosts of 1300 concurrent connections. I measured locally that one line is about 150 Byte. So, 1300x150 Byte = 19.5 KByte =~ 19 Kibibyte. 

If we play a bit with the numbers and say we will use the number of possible ports (and forget about other limits like open files), which is limited to 16 bits for the source port field in a UDPv4 header, we could have 2^16 = 65536 lines for a server listening on a single IP.

--> So, with 150 Byte per line we theoretically have a filesize of 65536 * 150 Byte ~ 10 MByte

But lets have a look at some numbers available at the internet:
- Whatsapp (published 2012: https://web.archive.org/web/20140501234954/https://blog.whatsapp.com/196/1-million-is-so-2011) they were able to handle 2M TCP connections at this time. Just for the fun lets say 2M UDP connections will then create a /proc/net/udp file with size of 300 MByte
- MigratoryData WebSocket Server (published: 2013 https://mrotaru.wordpress.com/2013/06/20/12-million-concurrent-connections-with-migratorydata-websocket-server/) they can handle 12M concurrent connections. Even this is not UDP related and more socket descriptor magic, let's use this number and we have a filesize of 1.8 GByte...

Means, we have a filesize range from more realistic KBytes to unusual GBytes. As an engineer I would choose the middle and estimate a maximum of multiple MByte per IP.

Reply:
Thanks for the followup, good stuff.

I guess we should stick to this streaming reader for large file safety. What do you think @mdlayher?

Reply:
I think it's most reasonable to stick to a streaming read but would also encourage using an io.LimitReader when initializing the scanner and set some sane upper bound like 4GiB of file.